### PR TITLE
Check-in matching (EXPOSUREAPP-5694)

### PR DESCRIPTION
**Pull request by chiljamgossow**
Calculates time overlap of check-ins and trace time warnings

Tech spec https://github.com/corona-warn-app/cwa-app-tech-spec/blob/proposal/event-registration-mvp/docs/spec/event-registration-client.md#calculate-overlap-of-checkin-and-tracetimeintervalwarning

This is how the algorithm works:
1. split check-ins by midnight GMT (reason for this is that risk is calculated per day)
2. check for all check-ins and relevant warnings if there is a time overlap
3. keep the overlaps for further processing -> risk calculation follows in a different PR 

Deletes the guidHashBase64 from TraceLocationCheckInEntity because this db field's only purpose is facilitating check-in matching, but proved to be unhelpful. Replaced by an attribute in the CheckIn class.

cannot be tested yet with the test menu, due to missing download/check-ins
but has plenty unit tests

so please check the code

### CODE REVIEW

File: Corona-Warn-App/schemas/de.rki.coronawarnapp.eventregistration.storage.TraceLocationDatabase/1.json [hunk 1]
```diff
@@ -116,7 +110,7 @@
       },
       {
         "tableName": "traceLocations",
-        "createSql": "CREATE TABLE IF NOT EXISTS `${TABLE_NAME}` (`guid` TEXT NOT NULL, `version` INTEGER NOT NULL, `type` INTEGER NOT NULL, `description` TEXT NOT NULL, `address` TEXT NOT NULL, `startDate` TEXT, `endDate` TEXT, `defaultCheckInLengthInMinutes` INTEGER, `byteRepresentation` TEXT NOT NULL, `signature` TEXT NOT NULL, PRIMARY KEY(`guid`))",
```
Comment (chiljamgossow) [inline Corona-Warn-App/schemas/de.rki.coronawarnapp.eventregistration.storage.TraceLocationDatabase/1.json:119]: This deletion shouldn't be in here, was not intended. Will fix it.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/eventregistration/checkins/split/CheckInSplitter.kt [hunk 2]
```diff
@@ -14,12 +15,15 @@ import kotlin.math.ceil
...
 fun CheckIn.splitByMidnightUTC(): List<CheckIn> {
     Timber.d("Starting splitByMidnightUTC ...")
+    if (checkInStart.toLocalDate() == checkInEnd.toLocalDate()) return listOf(this)
     val startTimeSeconds = checkInStart.seconds
     val endTimeSeconds = checkInEnd.seconds
     val durationSecondsUTC = endTimeSeconds - startTimeSeconds.toMidnightUTC()
 
     Timber.i("durationSecondsUTC=$durationSecondsUTC")
 
+    if (durationSecondsUTC <= 0) return listOf(this)
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/eventregistration/checkins/split/CheckInSplitter.kt:25]: Can you add a unit test for this case?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInMatching.kt [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+): List<CheckInWarningOverlap> {
+    return warningPackage
+        .extractTraceTimeIntervalWarnings()
+        .flatMap { warning ->
+            checkIns
+                .mapNotNull {
+                    it.calculateOverlap(warning)
+                }
+        }
+}
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInMatching.kt:49]: Please add some logging here.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInWarningMatcher.kt [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    private val dispatcherProvider: DispatcherProvider
+) {
+    suspend fun execute(): List<CheckInWarningOverlap> {
+        val checkIns = checkInsRepository.allCheckIns.firstOrNull() ?: return emptyList()
+
+        val warningPackages = traceTimeIntervalWarningRepository.allWarningPackages.firstOrNull() ?: return emptyList()
+
+        val splitCheckIns = checkIns.flatMap { it.splitByMidnightUTC() }
+
+        return launchMatching(splitCheckIns, warningPackages, dispatcherProvider.IO)
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInWarningMatcher.kt:23]: Love the parallelization effort!

We might have to cap the parallel executions though.
Currently we launch one coroutine (possibly one thread due to `IO` dispatcher) per `warningPackage`, which may be 24*15, though capped at 64 by kotlin (afaik).  Parallel I/O (reading the packages) is usually the bottle neck on most devices, especially low end ones, so 64 parallel attempts to read the files may starve the system of I/O resources and could cause the device to lag during our background execution.

This is all theoretical though, we'd need to test this with an actual set of warning packages (@mlenkeit is something like that available?).

Maybe we should cap parallel execution at 2-4 for now? What do you think? Anything above 4 might have diminishing returns anyways :thinking:

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInMatching.kt [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+import de.rki.coronawarnapp.util.TimeAndDateExtensions.toLocalDate
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.async
+import kotlinx.coroutines.withContext
+import org.joda.time.Duration
+import org.joda.time.Instant
+import kotlin.coroutines.CoroutineContext
+
+suspend fun launchMatching(
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInMatching.kt:15]: I'd prefer a different name to prevent missunderstandings as this does not "launch" until `await` is called on it, devs unexperienced with coroutines may be caught off-guard.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInMatching.kt [hunk 6]
```diff
@@ -0,0 +1,75 @@
+package de.rki.coronawarnapp.presencetracing.risk
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInMatching.kt:1]: What about moving these global functions into a class so they can be mocked better when writing unit tests for `CheckInWarningMatcher`?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInMatching.kt [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        List<CheckIn>,
+        TraceTimeIntervalWarningPackage
+    ) -> Deferred<List<CheckInWarningOverlap>> =
+        { list, warningPackage ->
+            async {
+                findMatches(list, warningPackage)
+            }
+        }
+
+    return warningPackages.map { warningPackage ->
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInMatching.kt:30]: Please add a unit test that shows how this is expected to behave if there is an error during execution.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInMatching.kt [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+                    it.calculateOverlap(warning)
+                }
+        }
+}
+
+fun CheckIn.calculateOverlap(
+    warning: TraceWarning.TraceTimeIntervalWarning
+): CheckInWarningOverlap? {
+
+    if (warning.locationGuidHash != locationGuidHash) return null
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInMatching.kt:55]: I'd love some logging for the case where there is a match, does not have to be right here, but somewhere?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInWarningMatcher.kt [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+class CheckInWarningMatcher @Inject constructor(
+    private val checkInsRepository: CheckInRepository,
+    private val traceTimeIntervalWarningRepository: TraceTimeIntervalWarningRepository,
+    private val dispatcherProvider: DispatcherProvider
+) {
+    suspend fun execute(): List<CheckInWarningOverlap> {
+        val checkIns = checkInsRepository.allCheckIns.firstOrNull() ?: return emptyList()
+
+        val warningPackages = traceTimeIntervalWarningRepository.allWarningPackages.firstOrNull() ?: return emptyList()
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInWarningMatcher.kt:19]: Some logging for the early aborts please :smile:

File: Corona-Warn-App/src/test/java/de/rki/coronawarnapp/presencetracing/risk/OverlapTest.kt [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+import de.rki.coronawarnapp.eventregistration.checkins.CheckIn
+import de.rki.coronawarnapp.server.protocols.internal.pt.TraceWarning
+import de.rki.coronawarnapp.util.HashExtensions.toSHA256
+import io.kotest.matchers.shouldBe
+import okio.ByteString
+import org.joda.time.Duration
+import org.joda.time.Instant
+import org.junit.Test
+
+class OverlapTest {
```
Comment (d4rken) [inline Corona-Warn-App/src/test/java/de/rki/coronawarnapp/presencetracing/risk/OverlapTest.kt:13]: `: BaseTest()`

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInMatching.kt [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+internal suspend fun findMatches(
+    checkIns: List<CheckIn>,
+    warningPackage: TraceTimeIntervalWarningPackage
+): List<CheckInWarningOverlap> {
+    return warningPackage
+        .extractTraceTimeIntervalWarnings()
+        .flatMap { warning ->
+            checkIns
+                .mapNotNull {
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInMatching.kt:45]: The `mapNotNull` is a crucial part of the matching calculation, maybe a test for this?

**Review by d4rken — COMMENTED** [review]
Nice! Some comments, no real blockers.

I think this empty file is a leftover from refactoring:
* `Corona-Warn-App/src/test/java/de/rki/coronawarnapp/eventregistration/checkins/DefaultCheckInsTransformerTest.kt`

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/eventregistration/checkins/split/CheckInSplitter.kt (hunk 2 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/eventregistration/checkins/split/CheckInSplitter.kt]: There was actually a bug here, thanks

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInMatching.kt (hunk 7 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInMatching.kt:31]: Most likely, it would be a file reading issue. No idea what should happen there. @mlenkeit What should happen if a package cannot be unpacked/evaluated? Omit the package or fail the whole calculation?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInWarningMatcher.kt (hunk 9 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInWarningMatcher.kt:19]: I will add logging in the risk calculation PR as the logic here changes

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInMatching.kt (hunk 6 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInMatching.kt:1]: I actually do not want them to be mocked in the CheckInWarningMatcherTest to have a test that covers the whole unit of work

Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInMatching.kt:1]: Should we make them part of that class? Why are they global functions in an extra file?

Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInMatching.kt:1]: I love Kotlin for offering global functions 😃 I can easily write unit tests for all kinds of edge cases without having to mock anything. And I can call them easily from everywhere, e.g the test menu. I can of course move them into the same file.

Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInMatching.kt]: I agree that they are very useful, but we run the risk of littering the Android Studio auto completion with functions that are only relevant in specific circumstances.

Scoping functions so only those that are interesting are available at specific places in the code make it easier for new developers to work on something, find relevant functions.

It's may also have a negative impact on IDE performance if there many global functions, similar to overuse of unscoped extension functions.

It would not be a blocker for me if you feel strongly about this.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInMatching.kt (hunk 7 above)
Comment (mlenkeit) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInMatching.kt]: Abort the calculation, ideally show a white card on the home screen.

Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInMatching.kt]: I will fix it in the other PR

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInMatching.kt (hunk 6 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInMatching.kt]: I marked them a internal, only visible for testing (for now)

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInWarningMatcher.kt [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+                }
+        }
+}
+
+@VisibleForTesting(otherwise = PRIVATE)
+internal fun CheckIn.calculateOverlap(
+    warning: TraceWarning.TraceTimeIntervalWarning
+): CheckInWarningOverlap? {
+
+    if (warning.locationGuidHash != locationGuidHash) return null
```
Comment (harambasicluka) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInWarningMatcher.kt:97]: The techspec mentions to return `0`, not null. So is this a wanted behaviour here?

Comment (harambasicluka) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInWarningMatcher.kt:97]: https://github.com/corona-warn-app/cwa-app-tech-spec/blob/proposal/event-registration-mvp/docs/spec/event-registration-client.md#calculate-overlap-of-checkin-and-tracetimeintervalwarning

**Review by harambasicluka — DISMISSED** [review]
LGTM 🚀 

Just left one question.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInWarningMatcher.kt (hunk 12 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInWarningMatcher.kt:97]: Yes, nulls are skipped for further processing

**Review by d4rken — DISMISSED** [review]
lgtm

**Review by harambasicluka — APPROVED** [review]

**Review by d4rken — APPROVED** [review]
